Java17strictfp

JuliusNmn opened the pull request:
Added isStrict attribute to classes.
Edited behavior of isStrict attribute flag in Method. The attribute is ignored if Java >= 17.

JuliusNmn committed:
Aded isSealed attribute to java classes

JuliusNmn committed:
modified isstrict behavior of methods

JuliusNmn committed:
fix typo

errt commented on OPAL/br/src/main/scala/org/opalj/br/Method.scala:
```diff
@@ -468,6 +468,16 @@ final class Method private[br] (
+     * @return wether this class is defined as strict. Starting from Java 17, this is true by default.
+     */
+    override def isStrict: Boolean =
+        if (this.classFile.version.major > bi.Java17MajorVersion)
```
Should be >=

errt commented on OPAL/br/src/main/scala/org/opalj/br/Method.scala:
```diff
@@ -468,6 +468,16 @@ final class Method private[br] (
 
+    /**
+     *
+     * @return wether this class is defined as strict. Starting from Java 17, this is true by default.
```
Why adding this here instead of just doing it directly in JVMMethod?

JuliusNmn replied:
JVMMethod does not have access to a classfile

errt commented on OPAL/br/src/main/scala/org/opalj/br/ClassFile.scala:
```diff
@@ -327,6 +328,16 @@ final class ClassFile private (
+     * @return wether this class is defined as strict. Starting from Java 17, this is true by default.
+     */
+    def isStrict: Boolean =
+        if (this.version.major >= bi.Java17MajorVersion)
```
I think this was also true by default in some very early versions of Java, I think we should treat that here as well.

JuliusNmn replied:
strictfp was introduced in java 1.2. In earlier versions, the float precision to use was unspecified. Would it be correct to return true here?

errt replied:
Are you sure it was unspecified? The JVM spec says "Java 1.0 and 1.1 required strict evaluation of floating-point expressions." Do you have a reference where it says it was unspecified?

errt replied:
Did you check this?

JuliusNmn replied:
The annotation actually does not exist for classes, I have removed it.

errt commented on OPAL/br/src/main/scala/org/opalj/br/ClassFile.scala:
```diff
@@ -327,6 +328,16 @@ final class ClassFile private (
     def isFinal: Boolean = (ACC_FINAL.mask & accessFlags) != 0
 
+    /**
+     *
```
If you add scaladoc, please do not only use tags (like the @return), but provide a description because that is what is primarily used for the scaladoc generation.

errt reviewed (changes requested):
I requested some small changes, but otherwise it looks good to me. How about adding a warning when trying to write out a classfile with version >= 17 and strictfp flag set?

JuliusNmn committed:
typo fix

JuliusNmn committed:
javadoc fix

JuliusNmn committed:
Removed isSealed attribute from Class. The access flag only exists for methods:
If a class has a strictfp modifier all of its methods will have ACC_STRICT set (in supported class file versions).

JuliusNmn committed:
removed unused import

JuliusNmn committed:
added warning if atrictfp method modifier is written out on class file versions where it is ignored.

JuliusNmn committed:
fixed isStrict behavior for methods to accurately reflect the specification (strict evaluation was required for all float expressions before Java 1.2)

JuliusNmn committed:
formatting

errt commented on OPAL/bc/src/main/scala/org/opalj/bc/Assembler.scala:
```diff
@@ -752,6 +754,9 @@ object Assembler {
             writeShort(methods.size)
             methods foreach { m ⇒
                 serialize(m)
+                if  ((ACC_STRICT.mask & m.access_flags) != 0 && (classFile.major_version < 46) || (classFile.major_version > 60)) {
```
There's an issue here with the parentheses. Unless I'm mistaken, this will always print an error for classfile version > 60 because the && binds stronger than the ||. I guess the parentheses right before and after the || should be removed to fix this?

JuliusNmn committed:
fix parantheses error

JuliusNmn committed:
Merge remote-tracking branch 'ssh/java17strictfp' into java17strictfp

# Conflicts:
#	OPAL/bc/src/main/scala/org/opalj/bc/Assembler.scala

errt committed:
Merge branch 'develop' into java17strictfp

JuliusNmn committed:
fix formatting

errt committed:
Merge branch 'develop' into java17strictfp
